File: src/bin/liq_finder.rs
//! marginfi liquidatable-account finder (read-only, Stage 1 live test).
//!
//! Scans every MarginfiAccount in the main group, prices each position from the
//! on-chain Pyth oracle the protocol itself reads (PriceUpdateV2), computes
//! maintenance health, and lists who is liquidatable (+ the closest near-misses,
//! so we can see how tight the market is). No money moves.
//!
//! Usage: HELIUS_RPC=<https json-rpc url> cargo run --release --bin liq_finder
//!   [NEAR=20]  how many near-liquidation accounts to show

use arb_engine::liquidation::{
    self as liq, Bank, BankMap, MarginfiAccount, PriceMap,
};
use solana_pubkey::Pubkey;
use std::collections::{HashMap, HashSet};
use std::time::Duration;

const MARGINFI_PROGRAM: &str = "MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA";
const MARGINFI_GROUP: &str = "4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8";

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() {
                return Some(v);
            }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}

fn b64(data: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    let s = data.get(0)?.as_str()?;
    base64::engine::general_purpose::STANDARD.decode(s).ok()
}

/// Batch getMultipleAccounts (100 keys/call) → map pubkey → raw bytes.
fn get_multiple(endpoint: &str, keys: &[Pubkey]) -> HashMap<Pubkey, Vec<u8>> {
    let mut out = HashMap::new();
    for chunk in keys.chunks(100) {
        let strs: Vec<String> = chunk.iter().map(|k| k.to_string()).collect();
        let Some(v) = rpc(endpoint, serde_json::json!({
            "jsonrpc":"2.0","id":1,"method":"getMultipleAccounts",
            "params":[strs, {"encoding":"base64"}]
        })) else { continue };
        for (i, acc) in v["result"]["value"].as_array().into_iter().flatten().enumerate() {
            if let Some(bytes) = acc.get("data").and_then(b64) {
                out.insert(chunk[i], bytes);
            }
        }
        std::thread::sleep(Duration::from_millis(50));
    }
    out
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC")
        .or_else(|_| std::env::var("RPC_HTTP"))
        .expect("HELIUS_RPC (a getProgramAccounts-capable JSON-RPC url) in .env");
    let near_n: usize = std::env::var("NEAR").ok().and_then(|s| s.parse().ok()).unwrap_or(20);

    // 1) All MarginfiAccounts in the main group. dataSlice trims to the balances
    //    region (1736 B) so the payload is ~half. Server-side dataSize filter
    //    still guarantees these are full 2312-byte accounts.
    eprintln!("[finder] getProgramAccounts (group {}) …", &MARGINFI_GROUP[..8]);
    let resp = rpc(&endpoint, serde_json::json!({
        "jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[MARGINFI_PROGRAM, {
            "encoding":"base64",
            "dataSlice":{"offset":0,"length":1736},
            "filters":[
                {"dataSize": liq::MA_SIZE},
                {"memcmp":{"offset":8,"bytes":MARGINFI_GROUP}}
            ]
        }]
    }));
    let accounts_json = resp.as_ref().and_then(|v| v["result"].as_array()).cloned().unwrap_or_default();
    eprintln!("[finder] {} accounts in group", accounts_json.len());
    if accounts_json.is_empty() {
        eprintln!("[finder] nothing returned — check RPC supports getProgramAccounts + the group filter");
        return;
    }

    let accounts: Vec<MarginfiAccount> = accounts_json.iter()
        .filter_map(|e| e["account"]["data"].get(0).and_then(|_| b64(&e["account"]["data"])))
        .filter_map(|bytes| MarginfiAccount::decode(&bytes))
        .filter(|a| !a.balances.is_empty())
        .collect();
    let borrowers: Vec<&MarginfiAccount> = accounts.iter()
        .filter(|a| a.balances.iter().any(|b| b.liability_shares > 0.0))
        .collect();
    eprintln!("[finder] {} accounts with balances, {} with an open borrow", accounts.len(), borrowers.len());

    // 2) Fetch every referenced Bank.
    let bank_pks: Vec<Pubkey> = borrowers.iter()
        .flat_map(|a| a.balances.iter().map(|b| b.bank_pk))
        .collect::<HashSet<_>>().into_iter().collect();
    eprintln!("[finder] fetching {} banks …", bank_pks.len());
    let bank_raw = get_multiple(&endpoint, &bank_pks);
    let mut banks: BankMap = HashMap::new();
    let mut oracle_of: HashMap<Pubkey, Pubkey> = HashMap::new();
    for (pk, raw) in &bank_raw {
        if let Some(bank) = Bank::decode(raw) {
            oracle_of.insert(*pk, bank.oracle_key);
            banks.insert(*pk, bank);
        }
    }

    // 3) Price each bank from its on-chain Pyth oracle (PriceUpdateV2).
    let oracle_pks: Vec<Pubkey> = oracle_of.values().copied().collect::<HashSet<_>>().into_iter().collect();
    eprintln!("[finder] fetching {} oracle accounts …", oracle_pks.len());
    let oracle_raw = get_multiple(&endpoint, &oracle_pks);
    let mut oracle_price: HashMap<Pubkey, f64> = HashMap::new();
    for (pk, raw) in &oracle_raw {
        if let Some(usd) = liq::decode_oracle_price(raw) {
            oracle_price.insert(*pk, usd);
        }
    }
    let mut prices: PriceMap = HashMap::new();
    for (bank_pk, oracle_pk) in &oracle_of {
        if let Some(&p) = oracle_price.get(oracle_pk) {
            prices.insert(*bank_pk, p);
        }
    }
    eprintln!("[finder] priced {}/{} banks", prices.len(), banks.len());

    // Sanity: dump a few priced banks (eyeball USDC≈$1, SOL≈$82, …).
    eprintln!("[finder] price sanity (mint → USD):");
    for (pk, bank) in banks.iter().take(200) {
        if let Some(&p) = prices.get(pk) {
            eprintln!("    {}… (dec {}) = ${:.4}", &bank.mint.to_string()[..8], bank.mint_decimals, p);
        }
    }

    // Dust threshold: below this seizable collateral (USD) a liquidation can't
    // cover gas+priority, so it isn't a real opportunity.
    let min_collateral: f64 = std::env::var("MIN_COLLATERAL_USD").ok()
        .and_then(|s| s.parse().ok()).unwrap_or(10.0);

    // 4) Health — only for borrowers whose EVERY bank we could price. An
    //    account with any unpriced bank is "incomplete", not a signal.
    struct Scored<'a> { assets: f64, deficit: f64, ratio: f64, a: &'a MarginfiAccount }
    let mut scored: Vec<Scored> = Vec::new();
    let mut incomplete = 0usize;
    for a in &borrowers {
        let r = liq::maintenance_health(a, &banks, &prices);
        if r.missing > 0 { incomplete += 1; continue; }
        scored.push(Scored {
            assets: r.health.weighted_assets,
            deficit: r.health.value(),
            ratio: r.health.ratio(),
            a,
        });
    }

    println!("\n════ marginfi liquidatable finder ════");
    println!("borrowers scanned:        {}", borrowers.len());
    println!("fully priced (judgable):  {}", scored.len());
    println!("incomplete (unpriced bank, skipped): {}", incomplete);

    // Liquidatable with REAL collateral to seize, ranked by seizable value.
    let mut real: Vec<&Scored> = scored.iter()
        .filter(|s| s.deficit < 0.0 && s.assets >= min_collateral).collect();
    real.sort_by(|x, y| y.assets.partial_cmp(&x.assets).unwrap_or(std::cmp::Ordering::Equal));
    let dust = scored.iter().filter(|s| s.deficit < 0.0 && s.assets < min_collateral).count();

    println!("LIQUIDATABLE (collateral ≥ ${:.0}): {}   [+{} dust ignored]",
        min_collateral, real.len(), dust);
    for s in real.iter().take(50) {
        println!("  authority {}…  collateral=${:.2}  deficit={:+.2} USD  liab/asset={:.4}",
            &s.a.authority.to_string()[..8], s.assets, s.deficit, s.ratio);
    }

    // Per-bank breakdown of the largest liquidatable account — tells us whether
    // the collateral is liquid (real opportunity) or a stuck/illiquid token.
    if let Some(top) = real.first() {
        println!("\n── breakdown: {}… (largest liquidatable) ──", &top.a.authority.to_string()[..8]);
        for b in &top.a.balances {
            let Some(bank) = banks.get(&b.bank_pk) else {
                println!("  bank {}… UNPRICED", &b.bank_pk.to_string()[..8]); continue;
            };
            let price = prices.get(&b.bank_pk).copied().unwrap_or(f64::NAN);
            let scale = 10f64.powi(bank.mint_decimals as i32);
            if b.asset_shares > 0.0 {
                let ui = b.asset_shares * bank.asset_share_value / scale;
                println!("  COLLATERAL mint {}… {:.4} tok × ${:.4} = ${:.2}  (maint w {:.2}, tier via weights)",
                    &bank.mint.to_string()[..8], ui, price, ui * price, bank.asset_weight_maint);
            }
            if b.liability_shares > 0.0 {
                let ui = b.liability_shares * bank.liability_share_value / scale;
                println!("  BORROW     mint {}… {:.4} tok × ${:.4} = ${:.2}  (maint w {:.2})",
                    &bank.mint.to_string()[..8], ui, price, ui * price, bank.liability_weight_maint);
            }
        }
    }

    // Closest healthy accounts WITH real collateral — the ones worth monitoring.
    let mut near: Vec<&Scored> = scored.iter()
        .filter(|s| s.deficit >= 0.0 && s.assets >= min_collateral).collect();
    near.sort_by(|x, y| y.ratio.partial_cmp(&x.ratio).unwrap_or(std::cmp::Ordering::Equal));
    println!("\nclosest to liquidation (collateral ≥ ${:.0}, liab/asset→1.0):", min_collateral);
    for s in near.iter().take(near_n) {
        println!("  {}…  liab/asset={:.4}  collateral=${:.2}  buffer={:+.2} USD",
            &s.a.authority.to_string()[..8], s.ratio, s.assets, s.deficit);
    }
    println!();
}

File: src/bin/liqFinder.ts
// Port of src/bin/liq_finder.rs
//
// marginfi liquidatable-account finder (read-only, Stage 1 live test).
//
// Scans every MarginfiAccount in the main group, prices each position from the
// on-chain Pyth oracle the protocol itself reads (PriceUpdateV2), computes
// maintenance health, and lists who is liquidatable (+ the closest near-misses,
// so we can see how tight the market is). No money moves.
//
// Usage: HELIUS_RPC=<https json-rpc url> tsx src/bin/liqFinder.ts
//   [NEAR=20]  how many near-liquidation accounts to show

import 'dotenv/config';
import { PublicKey } from '@solana/web3.js';
import {
  type BankMap,
  decodeBank,
  decodeMarginfiAccount,
  decodeOraclePrice,
  MA_SIZE,
  maintenanceHealth,
  type MarginfiAccount,
  type PriceMap,
} from '../lib/liquidation.js';

const MARGINFI_PROGRAM = 'MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA';
const MARGINFI_GROUP = '4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      const v = await res.json();
      return v;
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

function b64(data: any): Buffer | undefined {
  const s = data?.[0];
  if (typeof s !== 'string') return undefined;
  try {
    return Buffer.from(s, 'base64');
  } catch {
    return undefined;
  }
}

/** Batch getMultipleAccounts (100 keys/call) → map pubkey (base58) → raw bytes. */
async function getMultiple(endpoint: string, keys: PublicKey[]): Promise<Map<string, Buffer>> {
  const out = new Map<string, Buffer>();
  for (let i = 0; i < keys.length; i += 100) {
    const chunk = keys.slice(i, i + 100);
    const strs = chunk.map((k) => k.toBase58());
    const v = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getMultipleAccounts',
      params: [strs, { encoding: 'base64' }],
    });
    if (v === undefined) continue;
    const arr = v?.result?.value;
    if (!Array.isArray(arr)) continue;
    for (let j = 0; j < arr.length; j++) {
      const bytes = arr[j]?.data !== undefined ? b64(arr[j].data) : undefined;
      if (bytes !== undefined) out.set(chunk[j].toBase58(), bytes);
    }
    await sleep(50);
  }
  return out;
}

interface Scored {
  assets: number;
  deficit: number;
  ratio: number;
  a: MarginfiAccount;
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (!endpoint) throw new Error('HELIUS_RPC (a getProgramAccounts-capable JSON-RPC url) in .env');
  const nearN = Number.parseInt(process.env.NEAR ?? '', 10) || 20;

  // 1) All MarginfiAccounts in the main group. dataSlice trims to the balances
  //    region (1736 B) so the payload is ~half. Server-side dataSize filter
  //    still guarantees these are full 2312-byte accounts.
  console.error(`[finder] getProgramAccounts (group ${MARGINFI_GROUP.slice(0, 8)}) …`);
  const resp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [
      MARGINFI_PROGRAM,
      {
        encoding: 'base64',
        dataSlice: { offset: 0, length: 1736 },
        filters: [{ dataSize: MA_SIZE }, { memcmp: { offset: 8, bytes: MARGINFI_GROUP } }],
      },
    ],
  });
  const accountsJson: any[] = Array.isArray(resp?.result) ? resp.result : [];
  console.error(`[finder] ${accountsJson.length} accounts in group`);
  if (accountsJson.length === 0) {
    console.error('[finder] nothing returned — check RPC supports getProgramAccounts + the group filter');
    return;
  }

  const accounts: MarginfiAccount[] = [];
  for (const e of accountsJson) {
    const bytes = e?.account?.data !== undefined ? b64(e.account.data) : undefined;
    if (bytes === undefined) continue;
    const acct = decodeMarginfiAccount(bytes);
    if (acct === null) continue;
    if (acct.balances.length === 0) continue;
    accounts.push(acct);
  }
  const borrowers = accounts.filter((a) => a.balances.some((b) => b.liabilityShares > 0.0));
  console.error(`[finder] ${accounts.length} accounts with balances, ${borrowers.length} with an open borrow`);

  // 2) Fetch every referenced Bank.
  const bankPkSet = new Set<string>();
  const bankPks: PublicKey[] = [];
  for (const a of borrowers) {
    for (const b of a.balances) {
      const key = b.bankPk.toBase58();
      if (!bankPkSet.has(key)) {
        bankPkSet.add(key);
        bankPks.push(b.bankPk);
      }
    }
  }
  console.error(`[finder] fetching ${bankPks.length} banks …`);
  const bankRaw = await getMultiple(endpoint, bankPks);
  const banks: BankMap = new Map();
  const oracleOf = new Map<string, PublicKey>();
  for (const [pkStr, raw] of bankRaw) {
    const bank = decodeBank(raw);
    if (bank !== null) {
      oracleOf.set(pkStr, bank.oracleKey);
      banks.set(pkStr, bank);
    }
  }

  // 3) Price each bank from its on-chain Pyth oracle (PriceUpdateV2).
  const oraclePkSet = new Set<string>();
  const oraclePks: PublicKey[] = [];
  for (const oc of oracleOf.values()) {
    const key = oc.toBase58();
    if (!oraclePkSet.has(key)) {
      oraclePkSet.add(key);
      oraclePks.push(oc);
    }
  }
  console.error(`[finder] fetching ${oraclePks.length} oracle accounts …`);
  const oracleRaw = await getMultiple(endpoint, oraclePks);
  const oraclePrice = new Map<string, number>();
  for (const [pkStr, raw] of oracleRaw) {
    const usd = decodeOraclePrice(raw);
    if (usd !== null) oraclePrice.set(pkStr, usd);
  }
  const prices: PriceMap = new Map();
  for (const [bankPk, oraclePk] of oracleOf) {
    const p = oraclePrice.get(oraclePk.toBase58());
    if (p !== undefined) prices.set(bankPk, p);
  }
  console.error(`[finder] priced ${prices.size}/${banks.size} banks`);

  // Sanity: dump a few priced banks (eyeball USDC≈$1, SOL≈$82, …).
  console.error('[finder] price sanity (mint → USD):');
  let dumped = 0;
  for (const [pkStr, bank] of banks) {
    if (dumped >= 200) break;
    dumped += 1;
    const p = prices.get(pkStr);
    if (p !== undefined) {
      console.error(`    ${bank.mint.toBase58().slice(0, 8)}… (dec ${bank.mintDecimals}) = $${p.toFixed(4)}`);
    }
  }

  // Dust threshold: below this seizable collateral (USD) a liquidation can't
  // cover gas+priority, so it isn't a real opportunity.
  const minCollateral = Number.parseFloat(process.env.MIN_COLLATERAL_USD ?? '') || 10.0;

  // 4) Health — only for borrowers whose EVERY bank we could price. An
  //    account with any unpriced bank is "incomplete", not a signal.
  const scored: Scored[] = [];
  let incomplete = 0;
  for (const a of borrowers) {
    const r = maintenanceHealth(a, banks, prices);
    if (r.missing > 0) {
      incomplete += 1;
      continue;
    }
    scored.push({
      assets: r.health.weightedAssets,
      deficit: r.health.weightedAssets - r.health.weightedLiabilities,
      ratio: r.health.weightedAssets === 0.0 ? Infinity : r.health.weightedLiabilities / r.health.weightedAssets,
      a,
    });
  }

  console.log('\n════ marginfi liquidatable finder ════');
  console.log(`borrowers scanned:        ${borrowers.length}`);
  console.log(`fully priced (judgable):  ${scored.length}`);
  console.log(`incomplete (unpriced bank, skipped): ${incomplete}`);

  // Liquidatable with REAL collateral to seize, ranked by seizable value.
  const real = scored.filter((s) => s.deficit < 0.0 && s.assets >= minCollateral);
  real.sort((x, y) => y.assets - x.assets);
  const dust = scored.filter((s) => s.deficit < 0.0 && s.assets < minCollateral).length;

  console.log(
    `LIQUIDATABLE (collateral ≥ $${minCollateral.toFixed(0)}): ${real.length}   [+${dust} dust ignored]`,
  );
  for (const s of real.slice(0, 50)) {
    console.log(
      `  authority ${s.a.authority.toBase58().slice(0, 8)}…  collateral=$${s.assets.toFixed(2)}  deficit=${s.deficit >= 0 ? '+' : ''}${s.deficit.toFixed(2)} USD  liab/asset=${s.ratio.toFixed(4)}`,
    );
  }

  // Per-bank breakdown of the largest liquidatable account — tells us whether
  // the collateral is liquid (real opportunity) or a stuck/illiquid token.
  if (real.length > 0) {
    const top = real[0];
    console.log(`\n── breakdown: ${top.a.authority.toBase58().slice(0, 8)}… (largest liquidatable) ──`);
    for (const b of top.a.balances) {
      const bank = banks.get(b.bankPk.toBase58());
      if (bank === undefined) {
        console.log(`  bank ${b.bankPk.toBase58().slice(0, 8)}… UNPRICED`);
        continue;
      }
      const price = prices.get(b.bankPk.toBase58()) ?? NaN;
      const scale = 10 ** bank.mintDecimals;
      if (b.assetShares > 0.0) {
        const ui = (b.assetShares * bank.assetShareValue) / scale;
        console.log(
          `  COLLATERAL mint ${bank.mint.toBase58().slice(0, 8)}… ${ui.toFixed(4)} tok × $${price.toFixed(4)} = $${(ui * price).toFixed(2)}  (maint w ${bank.assetWeightMaint.toFixed(2)}, tier via weights)`,
        );
      }
      if (b.liabilityShares > 0.0) {
        const ui = (b.liabilityShares * bank.liabilityShareValue) / scale;
        console.log(
          `  BORROW     mint ${bank.mint.toBase58().slice(0, 8)}… ${ui.toFixed(4)} tok × $${price.toFixed(4)} = $${(ui * price).toFixed(2)}  (maint w ${bank.liabilityWeightMaint.toFixed(2)})`,
        );
      }
    }
  }

  // Closest healthy accounts WITH real collateral — the ones worth monitoring.
  const near = scored.filter((s) => s.deficit >= 0.0 && s.assets >= minCollateral);
  near.sort((x, y) => y.ratio - x.ratio);
  console.log(`\nclosest to liquidation (collateral ≥ $${minCollateral.toFixed(0)}, liab/asset→1.0):`);
  for (const s of near.slice(0, nearN)) {
    console.log(
      `  ${s.a.authority.toBase58().slice(0, 8)}…  liab/asset=${s.ratio.toFixed(4)}  collateral=$${s.assets.toFixed(2)}  buffer=${s.deficit >= 0 ? '+' : ''}${s.deficit.toFixed(2)} USD`,
    );
  }
  console.log();
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
